src/api/app/models/staging/staged_requests.rb

```diff
@@ -195,4 +172,32 @@ def link_xml(opts = {})
     # "<link package=\"foo\" project=\"bar\" rev=\"XXX\" cicount=\"copy\"/>"
     Nokogiri::XML::Builder.new { |x| x.link(opts) }.doc.root.to_s
   end
+
+  def stage_request(request)
+    bs_request_action = request.bs_request_actions.first
+    if bs_request_action.is_submit?
+      return unless package_can_be_linked?(request, bs_request_action.target_package)
+
+      link_package(bs_request_action)
+    end
+
+    # TODO: implement delete requests
+    ProjectLogEntry.create!(
+      project: staging_project,
+      user_name: user_login,
+      bs_request: request,
+      event_type: :staged_request,
+      datetime: Time.now,
+      package_name: bs_request_action.target_package
+    )
+    staging_project.staged_requests << request
+    add_review_for_staged_request(request)
+  end
+
+  def package_can_be_linked?(request, target_package)
+    return true unless Package.find_by(project: staging_project, name: target_package)
+
+    errors << "Can't stage request '#{request.number}': package '#{target_package}' already exists in '#{staging_project}'."
```

Comment:
This is changing the state of the error from inside a method that looks like a checker method... I would put the error filling part outside. You know, I could call `package_can_be_linked?` several times and this would be changing the object state...